pkg/tests/ossm/ior_test.go

```diff
@@ -0,0 +1,321 @@
+			createSimpleGateway(t)
+			checkSimpleGateway(t)
+
+			t.LogStepf("Record the Route before the upgrade")
+			before := getRoutes(t, meshNamespace)
+
+			t.LogStep("Upgrade SMCP to v2.4")
+			updateToV24SMCP(t, meshNamespace, meshName)
+
+			t.LogStep("Check the Routes existed afther the upgrade")
+			checkSimpleGateway(t)
+			after := getRoutes(t, meshNamespace)
+
+			t.LogStep("Check the Routes unchanged afther the upgrade")
+			if before[0].Metadata.ResourceVersion != after[0].Metadata.ResourceVersion {
+				t.Fatal("Expect the route to be unchanged, but it is changed after the upgrade")
+			} else {
+				t.LogSuccess("Got the same resourceVersion before and after the upgrade")
+			}
+		})
+
+		t.NewSubTest("check IOR does not delete routes after deleting Istio pod").Run(func(t test.TestHelper) {
+			total := 3
+			nsNames := []string{}
+			gateways := []string{}
+
+			t.Cleanup(func() {
+				if env.GetSMCPVersion().GreaterThanOrEqual(version.SMCP_2_4) {
+					removeIORCustomSetting(t, meshNamespace, meshName)
+				}
+
+				oc.DeleteNamespace(t, nsNames...)
+				oc.ApplyString(t, meshNamespace, GetSMMRTemplate())
+				oc.WaitSMMRReady(t, meshNamespace)
+			})
+
+			t.LogStep("Ensure the IOR enabled")
+			if env.GetSMCPVersion().GreaterThanOrEqual(version.SMCP_2_4) {
+				enableIOR(t, meshNamespace, meshName)
+			}
+
+			t.LogStepf("Create %d Gateways and they are in their own Namespace", total)
+			for i := 0; i < total; i++ {
+				ns := fmt.Sprintf("ns-%d", i)
+				nsNames = append(nsNames, ns)
+				gateways = append(gateways, generateGateway(fmt.Sprintf("gw-%d", i), ns, fmt.Sprintf("www-%d.test.ocp", i)))
+			}
+			oc.CreateNamespace(t, nsNames...)
+			oc.ApplyString(t, "", gateways...)
+
+			t.LogStepf("Update SMMR to include %d Namespaces", total)
+			start := time.Now()
+			oc.ApplyString(t, meshNamespace, fmt.Sprintf(`
+apiVersion: maistra.io/v1
+kind: ServiceMeshMemberRoll
+metadata:
+  name: default
+spec:
+  members:
+  - bookinfo
+  - foo
+  - bar
+  - legacy
+  - %s
+  `, strings.Join(nsNames, "\n  - ")))
+
+			var end time.Time
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				routes := getRoutes(t, meshNamespace)
+				if len(routes) == total {
+					end = time.Now()
+					t.LogSuccessf("Found all %d Routes in %.2f", total, end.Sub(start).Seconds())
+				} else {
+					t.Fatalf("Expect to find %d Routes but found %d instead", total, len(routes))
+				}
+			})
+
+			before := getRoutes(t, meshNamespace)
+			detectRouteChanges := func() {
+				retry.UntilSuccess(t, func(t test.TestHelper) {
+					after := getRoutes(t, meshNamespace)
+
+					if len(after) != total {
+						t.Fatalf("Expect %d Routes, but got %d instead", total, len(after))
+					}
+
+					if fmt.Sprint(buildRouteMap(before)) != fmt.Sprint(buildRouteMap(after)) {
+						t.Fatalf("Expect %d Routes remain unchanged, but they changed\nBefore: %v\nAfter: %v", total, before, after)
+					}
```

Comment:
When this fails, it's really hard to see what's wrong:
```
    ior_test.go:195:    FAILURE: Expect 3 Routes remain unchanged, but they changed
        Before: [{{19b1ebd8-02a1-466c-ad60-01037f4d65e4 ns-0-gw-0-b3de2ff7f5c56cf4 349620 map[maistra.io/gateway-name:gw-0 maistra.io/gateway-namespace:ns-0 maistra.io/gateway-resourceVersion:349332 maistra.io/generated-by:ior] map[maistra.io/original-host:www-0.test.ocp]}} {{492ca0fe-9f79-4329-947a-cf81b028aeae ns-1-gw-1-3df72d3a81944086 349621 map[maistra.io/gateway-name:gw-1 maistra.io/gateway-namespace:ns-1 maistra.io/gateway-resourceVersion:349335 maistra.io/generated-by:ior] map[maistra.io/original-host:www-1.test.ocp]}} {{6b05fdd1-95cb-4ca4-b778-82d407aef239 ns-2-gw-2-1968729e121634b3 349623 map[maistra.io/gateway-name:gw-2 maistra.io/gateway-namespace:ns-2 maistra.io/gateway-resourceVersion:349338 maistra.io/generated-by:ior] map[maistra.io/original-host:www-2.test.ocp]}}]
        After: [{{19b1ebd8-02a1-466c-ad60-01037f4d65e4 ns-0-gw-0-b3de2ff7f5c56cf4 350022 map[maistra.io/gateway-name:gw-0 maistra.io/gateway-namespace:ns-0 maistra.io/gateway-resourceVersion:349332 maistra.io/generated-by:ior] map[maistra.io/original-host:www-0.test.ocp]}} {{492ca0fe-9f79-4329-947a-cf81b028aeae ns-1-gw-1-3df72d3a81944086 349893 map[maistra.io/gateway-name:gw-1 maistra.io/gateway-namespace:ns-1 maistra.io/gateway-resourceVersion:349335 maistra.io/generated-by:ior] map[maistra.io/original-host:www-1.test.ocp]}} {{6b05fdd1-95cb-4ca4-b778-82d407aef239 ns-2-gw-2-1968729e121634b3 349894 map[maistra.io/gateway-name:gw-2 maistra.io/gateway-namespace:ns-2 maistra.io/gateway-resourceVersion:349338 maistra.io/generated-by:ior] map[maistra.io/original-host:www-2.test.ocp]}}]
```

Reply:
I think what is needed here is to wait a certain period of time to see if there is any changes to the routes. But once there is any change to the routes, we should fail immediately. We do not need retry here because it is not possible to un-change the route back what it was with the same `revision` number. 

The question is how long we should wait for after restarting the istiod pod or changing the `Gateway` ?